### PR TITLE
auth: reregister auth providers

### PR DESCRIPTION
**Pull request by mikedanese**
Alternatively we can register this in the mains of most (or all) components.

Fixes GKE e2es
Fixes: 5d721bff817ba4ef8c2cbab53bb46e69b955ebe3

That PR unregistered auth providers for kubectl and probably elsewhere.

```release-note
NONE
```

### CODE REVIEW

File: pkg/client/clientset_generated/internalclientset/clientset.go [hunk 1]
```diff
@@ -37,6 +37,9 @@ import (
 	schedulinginternalversion "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/scheduling/internalversion"
 	settingsinternalversion "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/settings/internalversion"
 	storageinternalversion "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/storage/internalversion"
+
+	// Import solely to initialize client auth plugins.
+	_ "k8s.io/client-go/plugin/pkg/client/auth"
```
Comment (lavalamp) [inline pkg/client/clientset_generated/internalclientset/clientset.go]: Import where it's needed, not here.

Comment (lavalamp) [reply pkg/client/clientset_generated/internalclientset/clientset.go]: So, yeah, in the mains of the components that need it.

Comment (liggitt) [reply pkg/client/clientset_generated/internalclientset/clientset.go]: Or from the code that constructs auth providers. Not sure we want non-uniform handling of auth providers from loaded kubeconfig files using client-go

**Comment by mikedanese** [comment]
Our options are in the mains like we have here or in "k8s.io/client-go/rest" which actually retrieves the transport.

**Comment by mikedanese** [comment]
@liggitt in "k8s.io/client-go/tools/clientcmd"?

**Comment by mikedanese** [comment]
@liggitt wants consistent handling of kubeconfig so he's recommending registering auth providers in clientcmd. Binaries can still construct raw rest.Configs and not import the auth providers.

**Comment by mikedanese** [comment]
Until gke goes green:
/hold

**Comment by mikedanese** [comment]
gke is green
/retest
/unhold

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by deads2k** [comment]
> All of those dependencies already exist in client-go, and were pulled down on clone, right? The listing in godep files is annoying, but I don't actually think it bothers me that much. @lavalamp and @deads2k can weigh in.

That depends on how your vendoring tool worked. Ones like glide walk packages back to find repo dependencies.  As an example, before this pull, depending on clientcmd would *not* pull in azure.  Now it would.  Pulling in azure brings with it the risk of conflicting levels of dependencies.  Because we (and many other libraries) leak dependent packages through their interfaces, you have to vendor with "strip-vendor".   Doing that with conflicting levels of Azure (or its transitives) produces a vendoring mess that can make it impossible to use the library.

The short version is that golang dependency management is a mess and this makes it noticeably worse.  I don't like any of the options.  Does anyone want to make a case for why someone who didn't care about the plugin should be forced to resolve conflicting level problems?

**Comment by liggitt** [comment]
> The short version is that golang dependency management is a mess and this makes it noticeably worse.

Ok. We can just register in kubectl for now. Long-term, the externalized exec auth provider will be recommended, and could conceivably be added by default, since it likely has a much smaller dependency impact.

> That PR unregistered auth providers for kubectl and probably elsewhere.

only kubectl packages imported that client, so I'd only expect that command to have been affected
